scene/gui/text_edit.cpp

```diff
@@ -2588,118 +2588,110 @@ void TextEdit::_do_backspace(bool p_word, bool p_all_to_left) {
-					if (words[c] < column) {
-						column = words[c];
+				for (int j = words.size() - 2; j >= 0; j -= 2) {
+					if (words[j] < prev_column) {
+						prev_column = words[j];
 						break;
 					}
 				}
 			}
 
-			// Check for any other carets in this range.
-			int overlapping_caret_index = -1;
+			// Remove overlapping carets.
 			for (int j = i + 1; j < caret_edit_order.size(); j++) {
-				// Check caret and selection in on the right line.
-				if (get_caret_line(caret_edit_order[j]) != get_caret_line(caret_idx) && (!has_selection(caret_edit_order[j]) || get_selection_to_line(caret_edit_order[j]) != get_caret_line(caret_idx))) {
+				int prev_caret = caret_edit_order[j];
+				// If a selection started from a different word and ended to this word, move deletion to the selection's beginning.
+				if (has_selection(prev_caret) && get_selection_to_line(prev_caret) == get_caret_line(caret_idx) && get_selection_to_column(prev_caret) > prev_column && (get_selection_from_line(prev_caret) < get_caret_line(caret_idx) || get_selection_from_column(prev_caret) <= prev_column)) {
+					prev_line = get_selection_from_line(prev_caret);
+					prev_column = get_selection_from_column(prev_caret);
+				} else if (get_caret_line(prev_caret) < prev_line || (get_caret_line(prev_caret) == prev_line && get_caret_column(prev_caret) <= prev_column)) {
+					// Outside of the word deleted by the current caret, so exit the loop.
 					break;
 				}
 
-				// If it has a selection, check it ends with in the range.
-				if ((has_selection(caret_edit_order[j]) && get_selection_to_column(caret_edit_order[j]) < column)) {
-					break;
-				}
-
-				// If it has a selection and it starts outside our word, we need to adjust the selection, and handle it later to prevent overlap.
-				if ((has_selection(caret_edit_order[j]) && get_selection_from_column(caret_edit_order[j]) < column)) {
-					carets.write[caret_edit_order[j]].selection.to_column = column;
-					overlapping_caret_index = caret_edit_order[j];
-					break;
-				}
-
-				// Otherwise we can remove it.
-				if (get_caret_column(caret_edit_order[j]) > column || (has_selection(caret_edit_order[j]) && get_selection_from_column(caret_edit_order[j]) > column)) {
-					deselect(caret_edit_order[j]);
-					carets_to_remove.push_back(caret_edit_order[j]);
-					set_caret_column(0, caret_idx == 0, caret_idx);
-					i = j;
-				}
+				carets_to_remove.push_back(caret_edit_order[j]);
+				i = j; // Don't update caret_idx, just mark the next caret from which to continue the operation.
 			}
 
-			_remove_text(get_caret_line(caret_idx), column, get_caret_line(caret_idx), from_column);
-
-			set_caret_line(get_caret_line(caret_idx), false, true, 0, caret_idx);
-			set_caret_column(column, caret_idx == 0, caret_idx);
-			adjust_carets_after_edit(caret_idx, get_caret_line(caret_idx), column, get_caret_line(caret_idx), from_column);
+		} else {
+			backspace();
+			end_action();
+			queue_redraw();
+			return;
```

Comment:
Not limiting this to a caret causes an extra backspace when using word delete in the following case:

https://user-images.githubusercontent.com/6584330/227723223-b3d29555-b717-4c8e-98f3-fadd0a566099.mp4